db/dcrpg/insightapi.go

```diff
@@ -0,0 +1,209 @@
+	copy(blockTransactions.STx, blockVerbose.STx)
+
+	return blockTransactions
+}
+
+// GetBlockHash returns the block hash by
+// the block height
+func (pgb *ChainDB) GetBlockHash(idx int64) (string, error) {
+	hash, err := RetrieveBlockHash(pgb.db, idx)
+	if err != nil {
+		log.Errorf("Unable to get block hash for block number %d: %v", idx, err)
+		return "", err
+	}
+	return hash, nil
+}
+
+// GetAddressBalance returns the balance of an address
+func (pgb *ChainDB) GetAddressBalance(address string, N, offset int64) *explorer.AddressBalance {
+	_, balance, err := pgb.AddressHistory(address, N, offset)
+	if err != nil {
+		return nil
+	}
+	return balance
+}
+
+// GetAddressInfo returns the basic info for an address
+func (pgb *ChainDB) GetAddressInfo(address string, N, offset int64) *apitypes.InsightAddressInfo {
+	rows, balance, err := pgb.AddressHistory(address, N, offset)
+	if err != nil {
+		return nil
+	}
+	var transactionIdList []string
+
+	var totalReceived, totalSent, unSpent dcrutil.Amount
+
+	totalReceived, _ = dcrutil.NewAmount(float64(balance.TotalSpent + balance.TotalUnspent))
+	totalSent, _ = dcrutil.NewAmount(float64(balance.TotalSpent))
+	unSpent, _ = dcrutil.NewAmount(float64(balance.TotalUnspent))
+
+	for _, row := range rows {
+		fundingTxId := row.FundingTxHash
+		if fundingTxId != "" {
+			transactionIdList = append(transactionIdList, fundingTxId)
+		}
+
+		spendingTxId := row.SpendingTxHash
+		if spendingTxId != "" {
+			transactionIdList = append(transactionIdList, spendingTxId)
+		}
+	}
+
+	addressInfo := &apitypes.InsightAddressInfo{
+		Address:        address,
+		TotalReceived:  totalReceived,
+		TransactionsID: transactionIdList,
+		TotalSent:      totalSent,
+		Unspent:        unSpent,
+	}
+	return addressInfo
+}
+
+// GetBlockSummaryTimeRange returns the blocks created within a specified time
+// range min, max time
+func (pgb *ChainDB) GetBlockSummaryTimeRange(min, max int64, limit int) []dbtypes.BlockDataBasic {
+	blockSummary, err := RetrieveBlockSummaryByTimeRange(pgb.db, min, max, limit)
+	if err != nil {
+		log.Errorf("Unable to retrieve block summary using time %d: %b", min, err)
+	}
+	return blockSummary
+}
+
+func makeAddressTxOutput(data *dcrjson.SearchRawTransactionsResult, address string) *apitypes.AddressTxnOutput {
+	tx := new(apitypes.AddressTxnOutput)
+	tx.Address = address
+	tx.TxnID = data.Txid
+	tx.Height = 0
+	tx.Confirmations = data.Confirmations
+
+	for i := range data.Vout {
+		if len(data.Vout[i].ScriptPubKey.Addresses) != 0 {
+			if data.Vout[i].ScriptPubKey.Addresses[0] == address {
+				tx.ScriptPubKey = data.Vout[i].ScriptPubKey.Hex
+				tx.Vout = data.Vout[i].N
+				tx.Atoms += data.Vout[i].Value
+			}
+		}
+	}
+
+	tx.Amount = tx.Atoms * 100000000
+	return tx
+}
+
+// GetAddressUTXO returns the utxo of an address
+func (db *ChainDB) GetAddressUTXO(address string) []*apitypes.AddressTxnOutput {
+	txs, err := rpcutils.SearchRawTransaction(db.client, address)
```

Comment:
How do we deal with the total and the count/skip limits?

Reply:
We should get the needed data from postgresql here, not RPC.  You'll have to look up the vouts for the input address, then get the rest of the info needed by whatever insight endpoint this is for.